alf/algorithms/merlin_algorithm.py

```diff
@@ -0,0 +1,441 @@
+            state=next_state,
+            info=LossInfo(
+                loss=decoder_loss.loss + kld,
+                extra=MBPLossInfo(decoder=decoder_loss, vae=kld)))
+
+
+class MemoryBasedActor(OnPolicyAlgorithm):
+    """The policy module for MERLIN model."""
+
+    def __init__(
+            self,
+            action_spec,
+            memory: MemoryWithUsage,
+            num_read_keys=1,
+            lstm_size=(100, 100),  # not sure what Merlin uses
+            latent_dim=20,
+            loss=None,
+            name="mba"):
+        """Create the policy module of MERLIN.
+
+        Args:
+            action_spec (nested BoundedTensorSpec): representing the actions.
+            memory (MemoryWithUsage): the memory module from MemoryBasedPredictor
+            num_read_keys (int): number of keys for reading memory.
+            latent_dim (int): the dimension of the hidden represenation of VAE.
+            lstm_size (list[int]): size of lstm layers
+            loss (None|ActorCriticLoss): an object for calculating the loss
+                for reinforcement learning. If None, a default ActorCriticLoss
+                will be used.
+            name (str): name of the algorithm.
+        """
+        # This is different from Merlin LSTM. This rnn only uses the output
+        # from ths last LSTM layer, while Merlin uses outputs from all LSTM
+        # layers
+        rnn = make_lstm_cell(lstm_size, name=name + "/lstm")
+
+        actor_input_dim = (
+            latent_dim + lstm_size[-1] + num_read_keys * memory.dim)
+
+        actor_net = ActorDistributionNetwork(
+            input_tensor_spec=TensorSpec((actor_input_dim, ),
+                                         dtype=tf.float32),
+            output_tensor_spec=action_spec,
+            fc_layer_params=(200, ),
+            activation_fn=tf.keras.activations.tanh,
+            name=name + "/actor_net")
+
+        super(MemoryBasedActor, self).__init__(
+            action_spec=action_spec,
+            train_state_spec=get_rnn_cell_state_spec(rnn),
+            action_distribution_spec=actor_net.output_spec)
+
+        self._loss = ActorCriticLoss(action_spec) if loss is None else loss
+        self._memory = memory
+
+        self._key_net = tf.keras.layers.Dense(
+            num_read_keys * (self._memory.dim + 1), name=name + "/key_net")
+
+        # TODO: add log p(a_i) as input to value net
+        value_input_dim = latent_dim
+        self._value_net = ValueNetwork(
+            input_tensor_spec=TensorSpec((value_input_dim, ),
+                                         dtype=tf.float32),
+            fc_layer_params=(200, ),
+            activation_fn=tf.keras.activations.tanh,
+            name=name + "/value_net")
+
+        self._rnn = rnn
+        self._actor_net = actor_net
+
+        # TODO: add qvalue_net for predicting Q-value
+
+    def _variables(self):
+        return _collect_variables(self._rnn, self._value_net, self._actor_net)
```

Comment:
_key_net should also be included.

Reply:
good catch. But since tf.keras.Layer now derives from tf.Module, we can get rid of _variables() now.